File: packages/ui-flutter/lib/src/dropdown_menu.dart
import 'package:flutter/material.dart';

import 'theme.dart';

/// Mirrors `packages/ui/src/components/dropdown-menu.tsx`. Wraps Flutter's
/// `MenuAnchor`, which renders items / separators / submenus and handles
/// dismissal — the "reuse the platform machinery" call. Put
/// `KinetixMenuItem` / `KinetixMenuSeparator` / `KinetixMenuLabel` (and
/// nested `SubmenuButton`s) in `menuChildren`; these row parts are shared
/// by KinetixContextMenu / KinetixMenubarMenu too.
class KinetixDropdownMenu extends StatelessWidget {
  const KinetixDropdownMenu({super.key, required this.menuChildren, required this.child});

  final List<Widget> menuChildren;
  final Widget child;

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    return MenuAnchor(
      style: MenuStyle(
        backgroundColor: WidgetStatePropertyAll(c.popover),
        surfaceTintColor: const WidgetStatePropertyAll(Color(0x00000000)),
        shape: WidgetStatePropertyAll(
          RoundedRectangleBorder(
            borderRadius: BorderRadius.circular(8),
            side: BorderSide(color: c.border),
          ),
        ),
      ),
      menuChildren: menuChildren,
      builder: (context, controller, _) => GestureDetector(
        behavior: HitTestBehavior.opaque,
        onTap: () => controller.isOpen ? controller.close() : controller.open(),
        child: child,
      ),
    );
  }
}

class KinetixMenuItem extends StatelessWidget {
  const KinetixMenuItem(
    this.label, {
    super.key,
    this.icon,
    this.destructive = false,
    required this.onPressed,
  });

  final String label;
  final IconData? icon;
  final bool destructive;
  final VoidCallback onPressed;

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    final Color color = destructive ? c.destructive : c.foreground;
    return MenuItemButton(
      onPressed: onPressed,
      leadingIcon: icon != null ? Icon(icon, size: 16, color: color) : null,
      child: Text(label, style: TextStyle(fontSize: 14, color: color)),
    );
  }
}

class KinetixMenuSeparator extends StatelessWidget {
  const KinetixMenuSeparator({super.key});

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    return Divider(height: 9, thickness: 1, color: c.border);
  }
}

class KinetixMenuLabel extends StatelessWidget {
  const KinetixMenuLabel(this.text, {super.key});

  final String text;

  @override
  Widget build(BuildContext context) {
    final c = KinetixTheme.of(context);
    return Padding(
      padding: const EdgeInsets.fromLTRB(12, 6, 12, 6),
      child: Text(
        text,
        style: TextStyle(fontSize: 11, fontWeight: FontWeight.w500, color: c.mutedForeground),
      ),
    );
  }
}

File: packages/ui-swiftui/Sources/KinetixUI/DropdownMenu.swift
//
// DropdownMenu.swift — KinetixDropdownMenu + shared menu-row parts.
//
// Mirrors packages/ui/src/components/dropdown-menu.tsx. Wraps SwiftUI's
// native `Menu`, which already renders items / separators / submenus and
// handles dismissal — the "reuse the platform machinery" call, same as
// the Compose port building on Material3's `DropdownMenu`. The Radix
// sub-component graph (Portal / Group / Sub / RadioGroup) isn't
// reproduced; place `KinetixMenuItem` / `KinetixMenuSeparator` /
// `KinetixMenuLabel` (and nested `Menu`s / `KinetixMenubarMenu` for
// submenus) directly in the `content` builder.
//
// These row parts are shared by KinetixContextMenu and KinetixMenubarMenu
// too — anywhere a native `Menu`/`.contextMenu` builder is expected.
//

import SwiftUI

public struct KinetixDropdownMenu<Label: View, Content: View>: View {
    private let menuContent: Content
    private let label: Label

    public init(
        @ViewBuilder content: () -> Content,
        @ViewBuilder label: () -> Label
    ) {
        self.menuContent = content()
        self.label = label()
    }

    public var body: some View {
        Menu {
            menuContent
        } label: {
            label
        }
    }
}

public struct KinetixMenuItem: View {
    private let title: String
    private let systemImage: String?
    private let role: ButtonRole?
    private let action: () -> Void

    public init(
        _ title: String,
        systemImage: String? = nil,
        destructive: Bool = false,
        action: @escaping () -> Void
    ) {
        self.title = title
        self.systemImage = systemImage
        self.role = destructive ? .destructive : nil
        self.action = action
    }

    public var body: some View {
        Button(role: role, action: action) {
            if let systemImage {
                Label(title, systemImage: systemImage)
            } else {
                Text(title)
            }
        }
    }
}

public struct KinetixMenuSeparator: View {
    public init() {}

    public var body: some View {
        Divider()
    }
}

public struct KinetixMenuLabel: View {
    private let text: String

    public init(_ text: String) {
        self.text = text
    }

    public var body: some View {
        Text(text)
    }
}
